vms/evm/subscriber.go

```diff
@@ -69,22 +117,150 @@ func NewWarpLogInfo(log types.Log) (*vmtypes.WarpLogInfo, error) {
 		SourceTxID:         log.TxHash[:],
 		UnsignedMsgBytes:   log.Data,
+		BlockNumber:        log.BlockNumber,
 	}, nil
 }
 
 // forward logs from the concrete log channel to the interface channel
 func (s *subscriber) forwardLogs() {
 	for msgLog := range s.evmLog {
-		messageInfo, err := NewWarpLogInfo(msgLog)
+		messageInfo, err := s.NewWarpLogInfo(msgLog)
 		if err != nil {
-			s.logger.Info(
+			s.logger.Error(
 				"Invalid log. Continuing.",
 				zap.Error(err),
 			)
 			continue
 		}
-		s.log <- *messageInfo
+		s.logsChan <- *messageInfo
+	}
+}
+
+// Process logs from the given block height to the latest block
+// Cap the number of blocks requested from the client to MaxBlocksToProcess,
+// counting back from the current block.
+func (s *subscriber) ProcessFromHeight(height *big.Int) error {
+	s.logger.Info(
+		"Processing historical logs",
+		zap.String("fromBlockHeight", height.String()),
+		zap.String("chainID", s.chainID.String()),
+	)
+	if height == nil {
+		return fmt.Errorf("cannot process logs from nil height")
+	}
+	ethClient, err := ethclient.Dial(s.nodeRPCURL)
+	if err != nil {
+		return err
+	}
+
+	// Grab the latest block before filtering logs so we don't miss any before updating the db
+	latestBlock, err := ethClient.BlockNumber(context.Background())
+	if err != nil {
+		s.logger.Error(
+			"Failed to get latest block",
+			zap.String("chainID", s.chainID.String()),
+			zap.Error(err),
+		)
+		return err
+	}
+
+	// Cap the number of blocks to process to MaxBlocksToProcess
+	toBlock := big.NewInt(0).SetUint64(latestBlock)
+	if height.Cmp(big.NewInt(0).Sub(toBlock, big.NewInt(MaxBlocksToProcess))) < 0 {
+		s.logger.Warn(
+			fmt.Sprintf("Requested to process too many blocks. Processing only the most recent %d blocks", MaxBlocksToProcess),
+			zap.String("requestedBlockHeight", height.String()),
+			zap.String("latestBlockHeight", toBlock.String()),
+			zap.String("chainID", s.chainID.String()),
+		)
+		height = big.NewInt(0).Add(toBlock, big.NewInt(-MaxBlocksToProcess))
+	}
+
+	// Filter logs from the latest seen block to the latest block
+	// Since initializationFilterQuery does not modify existing fields of warpFilterQuery,
+	// we can safely reuse warpFilterQuery with only a shallow copy
+	initializationFilterQuery := interfaces.FilterQuery{
+		Topics:    warpFilterQuery.Topics,
+		Addresses: warpFilterQuery.Addresses,
+		FromBlock: height,
+		ToBlock:   toBlock,
+	}
+	logs, err := ethClient.FilterLogs(context.Background(), initializationFilterQuery)
+	if err != nil {
+		s.logger.Error(
+			"Failed to get logs on initialization",
+			zap.String("chainID", s.chainID.String()),
+			zap.Error(err),
+		)
+		return err
+	}
+
+	// Sort the logs in ascending block order. Order logs by index within blocks
+	sort.SliceStable(logs, func(i, j int) bool {
+		if logs[i].BlockNumber == logs[j].BlockNumber {
+			return logs[i].TxIndex < logs[j].TxIndex
+		}
+		return logs[i].BlockNumber < logs[j].BlockNumber
+	})
```

Comment:
There can potentially be multiple logs in a single transaction. I think we need to handle that case by checking the log index.

Reply:
We should just be able to order by log index instead of tx index. I've made this change.